geo/src/algorithm/k_nearest_concave_hull.rs

```diff
@@ -0,0 +1,393 @@
+}
+
+impl<T> KNearestConcaveHull for Vec<Point<T>>
+where T: GeoFloat + RTreeNum
+{
+    type Scalar = T;
+    fn k_nearest_concave_hull(&self, k: u32) -> Polygon<Self::Scalar> {
+        concave_hull(self.clone(), k)
+    }
+}
+
+impl<T> KNearestConcaveHull for MultiPoint<T>
+    where T: GeoFloat + RTreeNum
+{
+    type Scalar = T;
+    fn k_nearest_concave_hull(&self, k: u32) -> Polygon<Self::Scalar> {
+        concave_hull(self.iter().map(|x| x.clone()).collect(), k)
+    }
+}
+
+impl<T> KNearestConcaveHull for LineString<T>
+where T: GeoFloat + RTreeNum
+{
+    type Scalar = T;
+    fn k_nearest_concave_hull(&self, k: u32) -> Polygon<Self::Scalar> {
+        concave_hull(self.points_iter().map(|x| x.clone()).collect(), k)
+    }
+}
+
+impl<T> KNearestConcaveHull for Polygon<T>
+    where
+        T: GeoFloat + RTreeNum,
+{
+    type Scalar = T;
+    fn k_nearest_concave_hull(&self, k: u32) -> Polygon<Self::Scalar> {
+        let points: Vec<Point<T>> = self.exterior().points_iter().map(|x| x.clone()).collect();
+        concave_hull(points, k)
+    }
+}
+
+impl<T> KNearestConcaveHull for MultiPolygon<T>
+    where
+        T: GeoFloat + RTreeNum,
+{
+    type Scalar = T;
+    fn k_nearest_concave_hull(&self, k: u32) -> Polygon<Self::Scalar> {
+        let points: Vec<Point<T>> = self.iter().flat_map(|poly| poly.exterior().points_iter().map(|x| x.clone())).collect();
+        concave_hull(points, k)
+    }
+}
+
+impl<T> KNearestConcaveHull for MultiLineString<T>
+    where
+        T: GeoFloat + RTreeNum,
+{
+    type Scalar = T;
+    fn k_nearest_concave_hull(&self, k: u32) -> Polygon<T> {
+        let points: Vec<Point<T>> = self.iter().flat_map(|poly| poly.points_iter().map(|x| x.clone())).collect();
+        concave_hull(points, k)
+    }
+}
+
+fn concave_hull<T>(points: Vec<Point<T>>, k: u32) -> Polygon<T>
+    where T: GeoFloat + RTreeNum
+{
+    let dataset = prepare_dataset(&points);
+    concave_hull_inner(dataset, k)
+}
+
+const DELTA: f32 = 0.000000001;
+
+// Removes duplicate points from the dataset.
+fn prepare_dataset<T>(points: &Vec<Point<T>>) -> rstar::RTree<Point<T>>
+    where T: GeoFloat + RTreeNum
+{
+    let mut dataset: rstar::RTree<Point<T>> = rstar::RTree::new();
+    for point in points {
+        let closest = dataset.nearest_neighbor(point);
+        if let Some(closest) = closest {
+            if taxicab_distance(closest, point) < T::from(DELTA).expect("Conversion from constant is always valid") {
+                continue;
+            }
+        }
+        
+        dataset.insert(point.clone())
+    }
+
+    dataset
+}
+
+fn taxicab_distance<T>(p1: &Point<T>, p2: &Point<T>) -> T
+    where T: GeoFloat + RTreeNum
+{
+    (p1.x() - p2.x()).abs() + (p2.y() - p2.y()).abs()
+}
+
+fn polygon_from_tree<T>(dataset: &rstar::RTree<Point<T>>) -> Polygon<T>
+    where T: GeoFloat + RTreeNum
+{
+    let mut points: Vec<Coordinate<T>> = dataset.iter().map(|p| p.0).collect();
+    points.push(points[0]);
```

Comment:
Upon further inspection... it seems like this method is only used when the original input has 3 or less points, so orderings here are all equivalent, but that's not obvious by reading this method.

So two alternatives I'd be OK with:
- do an `assert!(points.len() <= 3)` here to document this

or 

- delete this method and bail out earlier in the algorithm when we still have access to the initial point ordering rather than an RTree of those points.

Reply:
To use the second option we would have to change the algorithm signature to return Option<Polygon<T>>, so I prefer the first option.